### DISCUSSION TITLE
Comparison of radioactivedecay to PyNE and other codes

### DISCUSSION THREAD

**Discussion by vsisl** [radioactivedecay/radioactivedecay#91] [Q&A] [answered]
Hello!

First of all thank you for open-sourcing and maintaining this project! 

I am curious about the comparison of radiactivedecay to [PyNE](https://github.com/pyne/pyne) and other codes. I found the [radiactivedecay/comparisons](https://github.com/radioactivedecay/comparisons) repo but haven't tried running it yet. Could you share your thoughts about the advantages/disadvantages of radiactivedecay compared to other codes?

In particular I would like to know how it compares to PyNE in the ability to take into account meta-stable states. [PyNE has this open issue](https://github.com/pyne/pyne/issues/739) that prevents us from using it. Does radiactivedecay suffer from the same problem or we should be fine with it?

Besides that I would like to ask if you ever benchmarked radiactivedecay against other codes (besides PyNE and Radiological Toolbox). In particular, I'm curious about a comparison to [SCALE/ORIGEN](https://www.ornl.gov/scale/overview). 

Thank you for your help!

**Comment by alexmalins** [comment, marked as answer]
Hey there!

> Could you share your thoughts about the advantages/disadvantages of radiactivedecay compared to other codes?

> In particular I would like to know how it compares to PyNE in the ability to take into account meta-stable states. https://github.com/pyne/pyne/issues/739 that prevents us from using it. Does radiactivedecay suffer from the same problem or we should be fine with it?

TBH it's not really the code that determines whether meta-stable states are correctly taken into account, but rather it's how the decay chains are processed from the original dataset then fed into the code.

radioactivedecay uses the ICRP-107 dataset, which includes meta-stable states in its decay chains if they have a half-life of greater than about 1s: see more in my reply [here](https://github.com/radioactivedecay/radioactivedecay/issues/7#issuecomment-727160380). That's why we say radioactivedecay "supports metastable states".

PyNE has it's own processor for extracting decay chains from the ENSDF dataset, which doesn't correctly pull out metastable states that are within decay chains (PyNE issue you linked to).

One issue some people have with radioactivedecay is the ICRP-107 dataset only includes 1252 radionuclides. ENDF & ENSDF datasets contain many more. Whether this matters to you will be context dependent.


> Besides that I would like to ask if you ever benchmarked radiactivedecay against other codes (besides PyNE and Radiological Toolbox). In particular, I'm curious about a comparison to [SCALE/ORIGEN](https://www.ornl.gov/scale/overview).

I've not checked against anything else. If you do this yourself please let me know what you find!

SCALE/ORIGEN is the "industry standard" of decay codes - it's got a long history & advanced numerical routines. I imagine it will correctly deal with metastable states. However it is a convoluted process to obtain a copy that would cost me money personally, hence I've never tried it.

If you need metastable states and radioactivedecay/ICRP-107 doesn't have the decay chains you need, you could also check out [SandiaDecay](https://github.com/sandialabs/SandiaDecay). It's ENSDF based and correctly includes metastable states I believe.

Hope this helps!

**Reply by vsisl** [reply]
Thanks a lot for a quick reply and clarification!

> I've not checked against anything else. If you do this yourself please let me know what you find! 

We plan to make a benchmark against SCALE/ORIGEN. I was wondering that perhaps we could contribute to radioactivedecay by a pull-request with the benchmark results. If this is interesting for you and you have a preferred form for obtaining the results, let me know! :) I also need to check if the license allows sharing the SCALE results but I guess that should be no problem.

> One issue some people have with radioactivedecay is the ICRP-107 dataset only includes 1252 radionuclides. ENDF & ENSDF datasets contain many more.

I read somewhere that on your roadmap, you have enhancing support of radiactivedecay for other libraries such as ENDF. What is a status of this feature? Can we expect it any time soon?

Thanks again for all the tips and clarifications! It's very helpful.

**Reply by alexmalins** [reply]
> We plan to make a benchmark against SCALE/ORIGEN. I was wondering that perhaps we could contribute to radioactivedecay by a pull-request with the benchmark results. If this is interesting for you and you have a preferred form for obtaining the results, let me know! :) I also need to check if the license allows sharing the SCALE results but I guess that should be no problem.

Hi, yes you're welcome to submit what you find. There's a comparisons repo here: https://github.com/radioactivedecay/comparisons
For a fair comparison it makes sense to match the decay data between ORIGEN and radioactivedecay. This will probably require creating a decay dataset for the radioactivedecay with some of the ORIGEN data (e.g. a few of the decay chains that you are interested in...) 
Then any differences between the results of the codes can be attributed to their calculation & numerical accuracies.

> I read somewhere that on your roadmap, you have enhancing support of radiactivedecay for other libraries such as ENDF. What is a status of this feature? Can we expect it any time soon?

There's always a steady stream of requests so I've been meaning to take a week off work and put some development time into this project. The first things I'd do are switch the backend of the datasets to using pandas, as it's a much more usable way to store the data and allow for extensions such as adding decay radiations data. It would also make it easier for people to make their own datasets.

One of the things I'm not so motivated to do personally is the parsing of ENDF/ENSDF. Parsing those files is a lot of work as they are difficult file formats and have lots of edge cases & pain points with extracting chains. Especially with respect to correctly putting in metastable states ... which is why PyNE itself has had a problem for so long. If I was to do either, it would be ENSDF, which is actually a processed form of ENDF and so is easier to handle than ENDF.

Much easier and more likely than that is just to parse datasets that are already based on one of those libraries. E.g. SandiaDecay has an xml parsed version of ENSDF which is a much easier to handle.

But yea, improving the structure of radioactivedecay to ease the handling of datasets and adding decay radiations from ICRP-107 is higher on the priority list for me than adding other datasets. ICRP-107, although a little old now (it's based on ENSDF 2001 IIRC), is a fabulous dataset for many technical reasons, and one that's still suitable for a lot of problems in my view.
